File: python/src/mdpdesigner/export/explicit.py
"""PRISM/Storm explicit export (FORMAT.md §7), in the layout of PRISM's -exportmodel:
.tra transitions, .lab labels, and per reward structure .srew (state) and .trew (transition).

Mirrors web/src/io/export/explicit.ts.
"""

from __future__ import annotations

from ..model import Document
from ._common import (
    ExportFile,
    ExportResult,
    ExportWarning,
    branch_rewards,
    decimal,
    index_model,
    is_point_mass,
    reward_structures,
)


def export_explicit(doc: Document) -> ExportResult:
    m = doc.model
    ix = index_model(m)
    warnings: list[ExportWarning] = []
    mdp = m.type == "mdp"
    n = len(ix.state_ids)
    transitions = sum(len(row.branches) for row in ix.rows)

    # Local choice index k of each row within its state.
    local: list[int] = []
    for i, row in enumerate(ix.rows):
        same = i > 0 and ix.rows[i - 1].state == row.state
        local.append(local[i - 1] + 1 if same else 0)

    def src(i: int) -> str:
        return f"{ix.rows[i].state} {local[i]}" if mdp else f"{ix.rows[i].state}"

    def header(count: int) -> str:
        return f"{n} {len(ix.rows)} {count}" if mdp else f"{n} {count}"

    tra = [header(transitions)]
    for i, row in enumerate(ix.rows):
        has_action = mdp and row.choice is not None and row.choice.action is not None
        action = f" {row.choice.action}" if has_action else ""
        tra += [f"{src(i)} {t} {decimal(p)}{action}" for t, p in row.branches]

    if not is_point_mass(ix):
        warnings.append(
            ExportWarning(
                "The explicit format has no initial distributions: every state in the "
                'support of model.initial is labelled "init"'
            )
        )
    deadlocks = [row.state for row in ix.rows if row.choice is None]
    labels = [("init", [s for s, _ in ix.initial]), ("deadlock", deadlocks), *ix.labels]
    per_state: list[list[int]] = [[] for _ in ix.state_ids]
    for j, (_, states) in enumerate(labels):
        for s in states:
            per_state[s].append(j)
    lab = [" ".join(f'{j}="{name}"' for j, (name, _) in enumerate(labels))]
    lab += [f"{s}: {' '.join(map(str, js))}" for s, js in enumerate(per_state) if js]

    files = [
        ExportFile(".tra", "\n".join(tra) + "\n"),
        ExportFile(".lab", "\n".join(lab) + "\n"),
    ]
    for r in reward_structures(ix):
        title = f'# Reward structure "{r.id}"'
        if r.state:
            lines = [f"{s} {decimal(v)}" for s, v in r.state.items()]
            content = [title, f"{n} {len(lines)}", *lines]
            files.append(ExportFile(f".{r.id}.srew", "\n".join(content) + "\n"))
        lines = [
            f"{src(i)} {row.branches[j][0]} {decimal(v)}"
            for i, row in enumerate(ix.rows)
            for j, v in enumerate(branch_rewards(r, row))
            if v != 0
        ]
        if lines:
            content = [title, header(len(lines)), *lines]
            files.append(ExportFile(f".{r.id}.trew", "\n".join(content) + "\n"))
    return ExportResult(files, warnings)

File: web/src/io/export/explicit.ts
// PRISM/Storm explicit export (FORMAT.md §7), in the layout of PRISM's -exportmodel:
// .tra transitions, .lab labels, and per reward structure .srew (state) and .trew (transition).
import type { MdpDocument } from '../../core/types'
import { branchRewards, decimal, indexModel, isPointMass, rewardStructures, type ExportFile, type ExportResult, type ExportWarning } from './common'

export function exportExplicit(doc: MdpDocument): ExportResult {
  const m = doc.model
  const ix = indexModel(m)
  const warnings: ExportWarning[] = []
  const mdp = m.type === 'mdp'
  const n = ix.stateIds.length
  const transitions = ix.rows.reduce((k, row) => k + row.branches.length, 0)

  // Local choice index k of each row within its state.
  const local: number[] = []
  ix.rows.forEach((row, i) => local.push(i > 0 && ix.rows[i - 1].state === row.state ? local[i - 1] + 1 : 0))
  const src = (i: number) => (mdp ? `${ix.rows[i].state} ${local[i]}` : `${ix.rows[i].state}`)
  const header = (count: number) => (mdp ? `${n} ${ix.rows.length} ${count}` : `${n} ${count}`)

  const tra = [header(transitions)]
  ix.rows.forEach((row, i) => {
    const action = mdp && row.choice?.action !== undefined ? ` ${row.choice.action}` : ''
    for (const b of row.branches) tra.push(`${src(i)} ${b.target} ${decimal(b.prob)}${action}`)
  })

  if (!isPointMass(ix)) warnings.push({ message: 'The explicit format has no initial distributions: every state in the support of model.initial is labelled "init"' })
  const deadlocks = ix.rows.filter((row) => !row.choice).map((row) => row.state)
  const labels = [
    { id: 'init', states: ix.initial.map((x) => x.state) },
    { id: 'deadlock', states: deadlocks },
    ...ix.labels,
  ]
  const perState: number[][] = ix.stateIds.map(() => [])
  labels.forEach((l, j) => l.states.forEach((s) => perState[s].push(j)))
  const lab = [labels.map((l, j) => `${j}="${l.id}"`).join(' ')]
  perState.forEach((js, s) => js.length && lab.push(`${s}: ${js.join(' ')}`))

  const files: ExportFile[] = [
    { suffix: '.tra', content: tra.join('\n') + '\n' },
    { suffix: '.lab', content: lab.join('\n') + '\n' },
  ]
  for (const r of rewardStructures(ix)) {
    const title = `# Reward structure "${r.id}"`
    if (r.state.size) {
      const lines = [...r.state].map(([s, v]) => `${s} ${decimal(v)}`)
      files.push({ suffix: `.${r.id}.srew`, content: [title, `${n} ${lines.length}`, ...lines].join('\n') + '\n' })
    }
    const lines: string[] = []
    ix.rows.forEach((row, i) =>
      branchRewards(r, row).forEach((v, j) => {
        if (v.n !== 0n) lines.push(`${src(i)} ${row.branches[j].target} ${decimal(v)}`)
      }),
    )
    if (lines.length) files.push({ suffix: `.${r.id}.trew`, content: [title, header(lines.length), ...lines].join('\n') + '\n' })
  }
  return { files, warnings }
}
